aspect/tools/src/com/google/idea/blaze/aspect/PackageParser.java

```diff
@@ -68,14 +70,61 @@ static PackageParserOptions parseArgs(String[] args) {
 
   private static final Pattern PACKAGE_PATTERN = Pattern.compile("^\\s*package\\s+([\\w\\.]+)");
 
+  private static boolean isWorkerMode(String [] args) {
+    return args.length == 1 && args[0].equals("--persistent_worker");
+  }
+
+  private static String[] getWorkRequestArgs(WorkRequest workRequest) {
+    return workRequest.getArgumentsList().toArray(new String[0]);
+  }
+
+  private static void parsePackagesAndWriteManifest(
+      PackageParser parser, PackageParserOptions options) throws Exception {
+    Preconditions.checkNotNull(options.outputManifest);
+
+    Map<ArtifactLocation, String> outputMap = parser.parsePackageStrings(options.sources);
+    parser.writeManifest(outputMap, options.outputManifest);
+  }
+
+  private static void runPersistentWorker(PackageParser parser) throws IOException {
+    while (true) {
+      WorkRequest workRequest = WorkRequest.parseDelimitedFrom(System.in);
+      if (workRequest == null) {
+        // parseDelimitedFrom returns null iff the stream is at EOF, thus we have no more work to
+        // do.
+        return;
+      }
+
+      WorkResponse.Builder builder =
+          WorkResponse.newBuilder().setRequestId(workRequest.getRequestId());
+
+      try {
+        PackageParserOptions workRequestOptions = parseArgs(getWorkRequestArgs(workRequest));
+
+        parsePackagesAndWriteManifest(parser, workRequestOptions);
+
+        builder.build().writeDelimitedTo(System.out);
+      } catch (Exception e) {
```

Comment:
I'm wondering whether we should catch all exceptions this way here. In case an exception is thrown, the loop will still run. Maybe we should allow continuation of the loop only when "parsePackagesAndWriteManifest" throws a parsing-specific exception, and for other errors break the loop. The BazelJavaBuilder is doing so

https://github.com/bazelbuild/bazel/blob/a4251eab6988d6cf4f5e35681fbe2c1b0abe48ef/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/BazelJavaBuilder.java#L45-L55

But I'm not sure, otherwise LGTM :+1:

Reply:
I wanted to reduce the risk that we crash hard for an issue that only affects the current action. I honestly don't know what is best here. If you'd like me to be more conservative that's fine - will update the PR.

Reply:
Personally I would prefer to follow the same way as `BazelJavaBuilder`. For recoverable exceptions we can continue, but for unexpected ones System.exit would be more transparent. We could also print a message telling the user it is possible to run it without persistent workers.